src/expected_value.py

```diff
@@ -0,0 +1,95 @@
+            else:
+                sell_token = swap.pair.token1.symbol
+                buy_token = swap.pair.token0.symbol
+                sell_amount = swap.amount1_in
+                buy_amount = swap.amount0_out
+                volume = swap.amount_usd
+            o = {
+                'sellToken': sell_token,
+                'buyToken': buy_token,
+                'sellAmount': sell_amount,
+                'buyAmount': buy_amount,
+                'volume': volume
+            }
+            block_number = int(swap_transaction.block_number)
+            if block_number not in swaps_by_block.keys():
+                swaps_by_block[block_number] = []
+            swaps_by_block[block_number].append(o)
+    return swaps_by_block
+
+if not use_cache:
+    swaps_by_block = get_uniswap_swaps()
+    #sorted(swaps_by_block.keys(), reverse=True)
+    with open("uniswap_swaps.pickled", "bw+") as f:
+        pickle.dump(sorted(swaps_by_block.keys(), reverse=True), f)
+else:
+    with open("uniswap_swaps.pickled", "br") as f:
+        swaps_by_block = pickle.load(f)
+
+#print(swaps_by_block)
+#print({frozenset([o['sellToken'], o['buyToken']]) for b in swaps_by_block.values() for o in b})
+
+print("Expected waiting time for opposite offer in block:")
+number_of_blocks_looking_forward = 50
+focus_pairs=[['WETH', 'USDT', 'DAI', 'USDC'],['WETH', 'DAI'], ['WETH', 'USDT'],['WETH', 'USDC'] , ['WETH', 'MKR'], ['WETH', 'GNO'], ['WETH', 'SUSHI'], ['WETH', 'SWRV'], ['WETH', 'WBTC'], ['WETH', 'YFI'], ['WETH', 'LINK'] ]
+sorted_blocks = sorted(swaps_by_block.keys(), reverse=True)
+for focus_pair in focus_pairs:
+    prob_opposite_offer = dict()
+    for k in range(number_of_blocks_looking_forward):
+        nr_blocks_with_at_least_one_opposite_offer =0
+        avg_volume = 0
+        nr_direct_trades = 0
+        for i in range(len(sorted_blocks) - k):
+            block_i = sorted_blocks[i]
+            buy_tokens_i = {o['buyToken'] for o in swaps_by_block.get(block_i, [])}
+            sell_tokens_j = {
+                o['sellToken']
+                for j in range(block_i, block_i + k + 1)
+                for o in swaps_by_block.get(j, [])
+            }
+            if focus_pair is not None:
+                if focus_pair[0] in buy_tokens_i and \
+                    (focus_pair[1] in sell_tokens_j or \
+                    (len(focus_pair) >= 3 and focus_pair[2] in sell_tokens_j)  or \
+                    (len(focus_pair) >= 4 and focus_pair[3] in sell_tokens_j)
+                    ):
+                    nr_blocks_with_at_least_one_opposite_offer += 1
+                
+        prob_opposite_offer[k] = nr_blocks_with_at_least_one_opposite_offer / (len(sorted_blocks) - k)
+
+    expected_waiting_time_for_user = 0
+    last_probability = 0
+    for k, v in prob_opposite_offer.items():
+        #print(k,v)
+        expected_waiting_time_for_user += k*(v-min(last_probability, v))
+        last_probability = v
```

Comment:
I'm not sure what is being computed here, but note only in recent python versions the dictionaries are supposed to be kept sorted by insertion order.

Reply:
I am trying to calculate the expected value: E(x)= x*p(x), where p(x) = v-last_v. This min is just due to some weird errors that I got.

But I guess you explained the errors :D Damn, I have no clue about python...